### PR TITLE
Add python virtualenv to manage py modules.

### PR DESCRIPTION
**Pull request by halfflat**
Implements #13.

* Don't use absolute paths to python executables in ns_python assignments in system scripts, to allow correct use of virtualenv version.
* Add new ns_ variable ns_pyenv_modules that can be amended in system scripts for determining module set up in venv.
* Add scipy and xarray to default virtualenv module list.
* Add explicit mpi4py 3.0 to virtualenv module list for daint to help work around MPI link issue.
* Add virtualenv building in install-local.sh, together with command line flag for managing this.
* Quote arguments forwarded in run_with_mpi functions.

### CODE REVIEW

File: install-local.sh [hunk 1]
```diff
@@ -144,6 +167,28 @@ ns_timestamp=$(date +%Y-%m-%ST%H:%M:%S%z | sed 's/[0-9][0-9]$/:&/')
...
+    msg "Initializing python virtual environment"
+    ns_pyvenv_opt=
+    if [ "$ns_pyvenv" == inherit ]; then
+	ns_pyvenv_opt=--system-site-packages
+    fi
+
+    msg "Installing python modules: $ns_pyvenv_modules"
+    (
+	exec >> "$ns_build_path/log_pyvenv" 2>&1
+	if python3 -m venv $ns_pyvenv_opt "$ns_pyvenv_path"; then
```
Comment (bcumming) [inline install-local.sh]: Should this be `${ns_python}`? `python3` might not be enough, e.g. `python3.7`.

File: install-local.sh [hunk 2]
```diff
@@ -144,6 +167,28 @@ ns_timestamp=$(date +%Y-%m-%ST%H:%M:%S%z | sed 's/[0-9][0-9]$/:&/')
...
+	    source "$ns_pyvenv_path/bin/activate"
+	    for pkg in $ns_pyvenv_modules; do
+	        pip install "$pkg"
```
Comment (bcumming) [inline install-local.sh:186]: On OS X with homebrew you need to use `pip3` explicitly. Again, we might need a`ns_pip` variable, with some testing in the environment setup to find pip for Python 3.

File: systems/daint-gpu.sh [hunk 3]
```diff
@@ -9,13 +9,14 @@ ns_sysname="daint-gpu"
 module load daint-gpu
 module load CMake
 
-module load cudatoolkit/9.2.148_3.19-6.0.7.1_2.1__g3d9acc8
```
Comment (bcumming) [inline systems/daint-gpu.sh]: Why  remove cudatoolkit here?

File: systems/daint-gpu.sh [hunk 4]
```diff
@@ -9,13 +9,14 @@ ns_sysname="daint-gpu"
...
+module load cray-python/3.6.5.1
+ns_python=python3
 module load cray-hdf5 cray-netcdf
 
 # PyExtensions is needed for cython, mpi4py and others.
 # It loads cray-python/3.6.5.1 which points python at version 3.6.1.1
 module load PyExtensions/3.6.5.1-CrayGNU-18.08
-ns_python=$(which python3)
+ns_python=python3
```
Comment (bcumming) [inline systems/daint-gpu.sh:16]: This is already set above on line 13.

File: systems/daint-mc.sh [hunk 5]
```diff
@@ -11,14 +11,16 @@ module load CMake
...
-# It loads cray-python/3.6.5.1
-module load PyExtensions/3.6.5.1-CrayGNU-18.08
-ns_python=$(which python3)
+module load cray-python/3.6.5.1
+ns_python=python3
 
 # load after python tools because easybuild...
 module swap gcc/7.3.0
 
+# add mpi4py to virtualenv build
```
Comment (bcumming) [inline systems/daint-mc.sh:20]: Should this also be added in `systems/daint-gpu.sh`?

**Review by bcumming — CHANGES_REQUESTED** [review]
Looking good.

There are some small issues to address before merging.

File: install-local.sh (hunk 1 above)
Comment (halfflat) [reply install-local.sh]: It should, thanks.

File: install-local.sh (hunk 2 above)
Comment (halfflat) [reply install-local.sh:186]: Can I confirm with you that that still holds within the virtual environment? In principle, it should automatically 'install' pip in the environment, and I'm hoping that that means that `pip` alone will work.

File: systems/daint-gpu.sh (hunk 3 above)
Comment (halfflat) [reply systems/daint-gpu.sh]: Cut and paste errors here; will rectify.

File: systems/daint-mc.sh (hunk 5 above)
Comment (halfflat) [reply systems/daint-mc.sh:20]: Yes.

**Comment by halfflat** [comment]
Update (fc9cd1baf7bfdb5a606a47a6e939afc788ff1e09) address above comments except for the pip3 change. I just want to verify pip vs pip3 within virtualenv in Homebrew.

**Pull request merged**